gitindex/index.go

```diff
@@ -403,36 +408,37 @@ func IndexGitRepo(opts Options) error {
 		if when := commit.Committer.When; when.After(opts.BuildOptions.RepositoryDescription.LatestCommitDate) {
 			opts.BuildOptions.RepositoryDescription.LatestCommitDate = when
 		}
+	}
 
-		tree, err := commit.Tree()
-		if err != nil {
-			return fmt.Errorf("commit.Tree: %w", err)
-		}
+	if opts.Incremental && opts.BuildOptions.IncrementalSkipIndexing() {
+		return nil
+	}
 
-		ig, err := newIgnoreMatcher(tree)
-		if err != nil {
-			return fmt.Errorf("newIgnoreMatcher: %w", err)
-		}
+	// branch => (path, sha1) => repo.
+	var repos map[fileKey]BlobLocation
 
-		files, subVersions, err := TreeToFiles(repo, tree, opts.BuildOptions.RepositoryDescription.URL, repoCache)
-		if err != nil {
-			return fmt.Errorf("TreeToFiles: %w", err)
-		}
-		for k, v := range files {
-			if ig.Match(k.Path) {
-				continue
-			}
-			repos[k] = v
-			branchMap[k] = append(branchMap[k], b)
+	// fileKey => branches
+	var branchMap map[fileKey][]string
+
+	// Branch => Repo => SHA1
+	var branchVersions map[string]map[string]plumbing.Hash
+
+	if opts.BuildOptions.IsDelta {
+		repos, branchMap, branchVersions, opts.BuildOptions.ChangedOrRemovedFiles, err = prepareDeltaBuildMetadata(opts, repo)
```

Comment:
It's still possible to have a delta build that _doesn't_ tombstone any documents (example: a commit that only adds a new file).  In that case, having an empty `[]string Tombstones` isn't enough to disambiguate between a normal and a delta build. 

---

If this is confusing, how about using an interface to represent a builder instead of a single concrete struct? 


```go 

type Builder interface {
	Add() 
    Finish()
}

type DeltaBuilder struct {
	// implements Builder
}

type NormalBuilder struct {
	// implements Builder
}
```

That would allow us to move the `IsDelta` check outside of the build options.

Reply:
Sorrying my comment was not specific enough, I was talking about `ChangedOrRemovedFiels`. I think `IsDelta` should be a build option. I prefer a single implementation with IsDelta rather than moving control flow. I think multiple impl make sense when they diverge significantly / we have IsDelta checks too often which is a smell for a better design.